### PR TITLE
Canvas group sets course copy fix

### PR DESCRIPTION
**Pull request by marcospri**
For:

- https://github.com/hypothesis/product-backlog/issues/1219

## Testing

- Edit [an assignment](https://hypothesis.instructure.com/courses/125/assignments/873/edit?return_to=https%3A%2F%2Fhypothesis.instructure.com%2Fcourses%2F125%2Fassignments%2F873) in the original course. Don't save it, just arrive to the stage that list the group sets available.

We do this to simulate the original assignment configuration that, a side effect, stored the existing group sets in the DB.

- Launch [a groups assignment in the copied course](https://hypothesis.instructure.com/courses/607/assignments/5818) 

- [Change the name of the group](https://hypothesis.instructure.com/courses/607/groups#) set to match/not match the name of the old one: `Test group set` (Three dots menu on the right)

- Reset the mapping in the new course: ```delete from grouping where lms_id = '095247085b9091079de17a873cf5b9214f38c670' ; ```

- Change back to a not matching name to see the modal again.

--- 
Screenshot of the modal:

Screenshot from 2024-04-24 15-55-11

### CODE REVIEW

File: lms/product/plugin/course_copy.py [hunk 1]
```diff
@@ -102,7 +102,7 @@ def find_matching_group_set_in_course(self, course, group_set_id):
         # We might have a record of this because we just called `grouping_plugin.get_group_sets` as the current user
         # or another user might have done it before for us.
         if new_group_set := self._course_service.find_group_set(
-            name=group_set.name, context_id=course.lms_id
```
Comment (marcospri) [inline lms/product/plugin/course_copy.py:105]: Service now returns a dict

File: lms/static/scripts/frontend_apps/components/LaunchErrorDialog.tsx [hunk 2]
```diff
@@ -287,23 +287,32 @@ export default function LaunchErrorDialog({
...
-              settings and select a new group set.
-            </b>
+            If this is a <p>copied course</p> the instructor needs to create a
+            new group set in this course that matches the name of the group set
+            in the old course.
+            {error?.details?.group_set_name && (
+              <>
+                {' '}
+                The group set name in the old course was{' '}
+                {error.details.group_set_name}
```
Comment (marcospri) [inline lms/static/scripts/frontend_apps/components/LaunchErrorDialog.tsx:313]:
```
lms/static/scripts/frontend_apps/components/LaunchErrorDialog.tsx(309,30): 
error TS2339: Property 'group_set_name' 
does not exist on type 'string | object'.
```


I reckon this type anno for this is: https://github.com/hypothesis/lms/blob/main/lms/static/scripts/frontend_apps/errors.ts#L43

Should I somehow assert this is the object part and ignore the other? How?

File: lms/models/grouping.py [hunk 3]
```diff
@@ -167,6 +168,17 @@ class MoodleGroup(Grouping):
     __mapper_args__ = {"polymorphic_identity": Grouping.Type.MOODLE_GROUP}
 
 
+class GroupSet(TypedDict):
```
Comment (marcospri) [inline lms/models/grouping.py:171]: We store these as dicts in course.extra, define their type as dict.

File: lms/product/canvas/_plugin/grouping.py [hunk 4]
```diff
@@ -78,8 +79,16 @@ def get_groups_for_instructor(self, _svc, course, group_set_id):
...
+            if group_set := self._request.find_service(name="course").find_group_set(
+                group_set_id=group_set_id
+            ):
+                group_set_name = group_set["name"]
+
             raise GroupError(
-                ErrorCodes.GROUP_SET_NOT_FOUND, group_set=group_set_id
+                ErrorCodes.GROUP_SET_NOT_FOUND,
+                group_set=group_set_id,
+                group_set_name=group_set_name,
```
Comment (marcospri) [inline lms/product/canvas/_plugin/grouping.py:91]: The name we are looking for is pretty important, more so in Canvas which doesn't copy the group sets so they need to be created by hand.

Expose the old name if we have it.

File: lms/static/scripts/frontend_apps/components/LaunchErrorDialog.tsx (hunk 2 above)
Comment (marcospri) [reply lms/static/scripts/frontend_apps/components/LaunchErrorDialog.tsx:313]: @acelaya This is not ready for review but adding you here to take a look at this

**Review by acelaya — APPROVED** [review]

**Pull request merged**